### ISSUE TITLE
Last.fm integration

### ISSUE THREAD

**Issue by guraltsev** [Mastermindzh/tidal-hifi#4]
Last.fm integration doesn't seem to work. I click on the button in "settings". A new window opens. I log into last.fm and authorize the music player and the window changes to all white with the message

```
This window may now be closed.
```

But then, once I close the window, if I go back into settings last.fm still seems disconnected. 

Signing into last.fm through chrome or firefox works fine. 

Please let me know if I can provide any more info. Thank you very much for the app!

**Comment by Mastermindzh** [comment]
Just writing in to let you know I've received this and I will start investigating it next week.

Sorry, been incredibly busy :)

**Comment by Mastermindzh** [comment]
I confirmed the bug, I don't have a clue how to fix it yet.

If anyone has an idea I'd love to hear about it.

**Comment by markkuit** [comment]
I ran thorough testing on this, and I'm positive it's due to [Electron lack of support for window.opener](https://github.com/electron/electron/issues/1865).
The last.fm popup window tries to send its token to Tidal's callback, but it can't access it due to said limitations.

> ➜ env ELECTRON_ENABLE_LOGGING=1 npm start
> 
> tidal-hifi@1.0.1 start /home/markkuit/dev/tidal-hifi
> electron .
> 
> [1822651:0428/150228.794579:INFO:CONSOLE(24)] "Uncaught TypeError: window.opener.receiveLastFMLoginMessage is not a function", source: https://listen.tidal.com/lastfmcallback.html?path=/settings&token=[...] (24)

I managed to get it working by enabling experimental support via the [sandbox](https://www.electronjs.org/docs/api/sandbox-option) option, which makes so that renderers must delegate certain tasks to the main process via IPC for security and isolation purposes, having Chromium render the content and incidentally working around the lack of support for `window.opener`. This still poses some limitations (see [this](https://github.com/electron/electron/issues/1865#issuecomment-249989894) and [this](https://github.com/electron/electron/issues/1865#issuecomment-249991896)), however none of those should matter, mainly because we don't care for Node.js in this scenario.

An alternative solution might have been using the [nativeWindowOpen](https://www.electronjs.org/docs/api/window-open#using-chromes-windowopen-implementation) option (as suggested [here](https://github.com/electron/electron/issues/1865#issuecomment-402425512)), which however caused me some headaches due to CSP/CORS policies and I ended up discarding it.

For the sake of tracking, I'll open a pull request for this. I didn't up the package version, that's up to you.

I'll take the occasion to thank you for this much appreciated project.

**Comment by Mastermindzh** [comment]
No problem mate, high fidelity music should be listened to on every platform.
Until Tidal decides to build and maintain a solution for all platforms I'll keep working on this project :)

**Comment by Mastermindzh** [comment]
The readme now has a workaround to "fix" this :)
Enjoy!

**Comment by Tuuuukka** [comment]
@Mastermindzh, what do you actually use as the Rescrobbled `player-whitelist` value to target Tidal-hifi? Using `playerctl --list-all` I get something like `chromium.instance2168528` for Tidal-hifi, but the numbers change every time it's restarted, so it's not a long term solution. 

Of course it works if I leave the `player-whitelist` empty, but then Rescrobbled would scrobble media from every possible player it recognizes, which is not good.

**Comment by Mastermindzh** [comment]
@Tuuuukka , I don't use last.fm, I'm pretty much relying entirely on others here, sorry 😄 

If you know/have any idea how one would change the name of that 'chromium.instance' (even for general linux apps) I might be able to figure something out.

**Comment by Tuuuukka** [comment]
>If you know/have any idea how one would change the name of that 'chromium.instance' (even for general linux apps) I might be able to figure something out.

Sorry, I don't know how that could be done (or if it's possible at all anyway 😄). That would of course be the clearest way to target Tidal-hifi in Rescrobbled.

**Comment by Mastermindzh** [comment]
I'll have a look next time I work on this (probably the weekend)

**Comment by hestonhoffman** [comment]
> @Mastermindzh, what do you actually use as the Rescrobbled `player-whitelist` value to target Tidal-hifi? Using `playerctl --list-all` I get something like `chromium.instance2168528` for Tidal-hifi, but the numbers change every time it's restarted, so it's not a long term solution.

In case anyone else stumbles across this and is looking for an answer, you can use this to whitelist tidal-hifi in Rescrobbled:
```toml
player-whitelist = [ "chromium" ]
```

**Comment by BlueManCZ** [comment]
I came across this issue and just wanted to report that I was able to connect to Last.fm. The external window maybe did not render correctly (maybe wayland issue?), but I run the application with `tidal-hifi --no-sandbox` for the first time to log in and since then the scobbling works just fine. If others will confirm this, maybe we can update the readme.

![obrazek](https://user-images.githubusercontent.com/17854950/190444329-92397601-9d9d-41b5-8638-756b30913cb2.png)
